### ISSUE TITLE
la génération du pdf plante avec message "context has timed out"

### ISSUE THREAD

**Issue by oliviercrouzet** [edinum/pdfgen#5]
Bonjour 
Après ajout des derniers patchs, je viens de tester sur 2 cas qui posaient problèmes à cause des images (pdf tronqués).
L'un fonctionne maintenant. Pour l'autre, ça plante avant même l'affichage du pdf avec le message suivant :
 ```
PHP Error  in file '/var/www/lodel/share/plugins/custom/pdfgen/vendor/thecodingmachine/gotenberg-php-client/src/Client.php' on line 106 : 
{"message":"context has timed out"}
```
l'url de l'article (voir les sources en fichier attaché)
https://publications-prairial.fr/cliothemis/index.php?id=568
[CLIO_15_06_chenu.zip](https://github.com/edinum/pdfgen/files/6636622/CLIO_15_06_chenu.zip)

**Comment by brrd** [comment]
Le template HTML s'affiche correctement : https://publications-prairial.fr/cliothemis/?do=_pdfgen_view&document=568&lang=fr

C'est apparemment un problème de timeout avec Gotenberg, actuellement il semble qu'on est à 10 (par défaut) : https://github.com/thecodingmachine/gotenberg/issues/144

Il faudrait monter le délai à une valeur importante (`waitTimeout = 30`) pour les articles longs ou avec beaucoup d'images (ici il y a 10 Mo d'images). Et aussi s'assurer que la page n'est pas mise en cache quand Gotenberg envoie cette erreur (mais j'ai pas l'impression car je mesure toujours un temps de réponse de 10s).

@nahuelange c'est plutôt ta partie non ?

**Comment by nahuelange** [comment]
Oui, je vais tenter de reproduire le problème chez nous pour fournir une solution.

**Comment by brrd** [comment]
Depuis la version 7, Gotenberg propose l'option waitForExpression qui permet d'exécuter la génération quand le JS le demande : https://gotenberg.dev/docs/routes#wait-before-rendering-chromium

> The JavaScript expression to wait before converting an HTML document into PDF until it returns true.

Actuellement on utilise Gotenberg 6 et waitTimeout ce qui nous oblige à d'attendre une durée arbitraire avant de la lancer pour être sûr que la page a fini de charger. Le souci c'est que plus les documents grossissent plus il faut augmenter cette durée pour tous les PDF (y compris les plus petits).

Upgrader Gotenberg et utiliser cette nouvelle feature permettrait de mieux gérer ça.

**Comment by brrd** [comment]
FTR suite de l'enquête : passer à Gotenberg 7+ implique de passer au client PHP https://github.com/gotenberg/gotenberg-php qui n'est compatible qu'avec PHP 8. On est donc bloqués tant que Lodel ne passe pas le cap de PHP 7.

**Comment by brrd** [comment]
L'erreur "context has timed out" n'apparaît plus. J'ouvre un nouveau ticket pour les erreurs actuelles de PDF incomplets : #23.
